### PR TITLE
Merge auth cookies

### PR DESCRIPTION
**Pull request by ckr123**

### CODE REVIEW

File: domains/auth/Auth.API/API/Controllers/LogoutController.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        this.tokenStorage = tokenStorage;
+        this.authOptions = authOptions.Value;
+        this.oidcService = oidcService;
+    }
+
+    [HttpPost]
+    [Route("/auth/logout")]
+    public async Task<ActionResult<LogoutResponse>> Logout()
+    {
+        var opaqueToken = HttpContext.Request.Headers[authOptions.CookieName].FirstOrDefault()?.Split(" ").Last();
```
Comment (C-Christiansen) [inline domains/auth/Auth.API/API/Controllers/LogoutController.cs:32]: var opaqueToken = HttpContext.Request.Cookies[authOptions.CookieName];

File: domains/auth/Auth.API/API/Controllers/TokenController.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+using API.Models;
+using API.Services;
+using API.Services.OidcProviders;
+using API.TokenStorage;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Options;
+
+namespace API.Controllers;
+
+
```
Comment (endk-awo) [inline domains/auth/Auth.API/API/Controllers/TokenController.cs:12]: Redundant newline

File: domains/auth/Auth.API/API/Controllers/TokenController.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+[ApiController]
+public class TokenController : ControllerBase
+{
+    readonly ITokenStorage tokenStorage;
```
Comment (endk-awo) [inline domains/auth/Auth.API/API/Controllers/TokenController.cs:16]: Unify accessibility modifiers (add private)

File: domains/auth/Auth.API/API/Controllers/TokenController.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        this.cookies = cookies;
+    }
+
+    [HttpGet]
+    [Route("/token/forward-auth")]
+    public ActionResult ForwardAuth()
+    {
+        var opaqueToken = HttpContext.Request.Cookies[authOptions.CookieName];
+
+        if (opaqueToken == null || opaqueToken == "")
```
Comment (endk-awo) [inline domains/auth/Auth.API/API/Controllers/TokenController.cs:33]: Use string.IsNullOrWhitespace(opaqueToken)

File: domains/auth/Auth.API/API/Controllers/LogoutController.cs [hunk 5]
```diff
@@ -28,8 +29,7 @@ IOidcService oidcService
     [Route("/auth/logout")]
     public async Task<ActionResult<LogoutResponse>> Logout()
     {
-        var opaqueToken = HttpContext.Request.Headers[authOptions.CookieName].FirstOrDefault()?.Split(" ").Last();
-
+        var opaqueToken = HttpContext.Request.Cookies[authOptions.CookieName];
         if (opaqueToken != null)
```
Comment (endk-awo) [inline domains/auth/Auth.API/API/Controllers/LogoutController.cs:33]: Check for whitespace in opaqueToken?

File: domains/auth/Auth.API/API/Repository/ITokenStorage.cs [hunk 6]
```diff
@@ -1,8 +1,12 @@
-namespace API.Services
+using API.Models;
+
+namespace API.TokenStorage
```
Comment (endk-awo) [inline domains/auth/Auth.API/API/Repository/ITokenStorage.cs:3]: Unify use of curly braces for namespaces

File: domains/auth/Auth.API/API/Repository/TokenStorage.cs [hunk 7]
```diff
@@ -1,4 +1,6 @@
-namespace API.Services
+using API.Models;
+
+namespace API.TokenStorage
```
Comment (endk-awo) [inline domains/auth/Auth.API/API/Repository/TokenStorage.cs:3]: Unify use of curly braces for namespaces

File: domains/auth/Auth.API/API/Repository/TokenStorage.cs [hunk 8]
```diff
@@ -11,5 +13,40 @@ public string GetIdTokenByOpaqueToken(string token)
...
+            }
+
+            return true;
+        }
+
+        public InternalToken? GetInteralTokenByOpaqueToken(string token)
+        {
+            // TODO Get interalToken from DB
+
+            var interalToken = new InternalToken();
```
Comment (endk-awo) [inline domains/auth/Auth.API/API/Repository/TokenStorage.cs:36]: Spelling: interalToken vs. internalToken

File: domains/auth/Auth.API/API/Repository/TokenStorage.cs [hunk 9]
```diff
@@ -11,5 +13,40 @@ public string GetIdTokenByOpaqueToken(string token)
...
+            if (interalToken == null)
+            {
+                return null;
+            }
+
+            var isValid = InternalTokenValidation(interalToken);
+
+            if (isValid != true)
```
Comment (endk-awo) [inline domains/auth/Auth.API/API/Repository/TokenStorage.cs:44]: Consider: 

if(!isValid) return null;

or even:

if(!InternalTokenValidation(internalToken)) return null;

File: domains/auth/Auth.API/API/Utilities/Cookies.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
+using API.Configuration;
+using Microsoft.Extensions.Options;
+
+
```
Comment (endk-awo) [inline domains/auth/Auth.API/API/Utilities/Cookies.cs:4]: Redundant newline

File: domains/auth/Auth.API/API/Utilities/ICookies.cs [hunk 11]
```diff
@@ -0,0 +1,5 @@
+namespace API.Services;
```
Comment (endk-awo) [inline domains/auth/Auth.API/API/Utilities/ICookies.cs:1]: Missing newline

File: domains/auth/Auth.API/Tests/Controllers/TestLogoutController.cs [hunk 12]
```diff
@@ -39,32 +44,23 @@ public TestLogoutController()
...
-            Domain = "energioprindelse.dk",
-            HttpOnly = true,
-            SameSite = SameSiteMode.Strict,
-            Secure = true,
-            Expires = DateTime.UtcNow.AddHours(6),
-        };
+    [Theory, MemberData(nameof(Cookie))]
+    public async Task LogoutDeleteCookieReturnSuccess(string? testCookie)
+    {
+        var expectedExpiredCookie = "Authorization=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/";
```
Comment (endk-awo) [inline domains/auth/Auth.API/Tests/Controllers/TestLogoutController.cs:56]: Consider using:

// Arrange
// Act
// Assert

File: domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    private readonly TokenController tokenController;
+
+    public TestTokenController()
+    {
+        authOptionsMock.Setup(x => x.Value).Returns(new AuthOptions
+        {
+            CookieName = "Authorization",
+        });
+
+        tokenController = new TokenController(
```
Comment (endk-awo) [inline domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:32]: This looks a little weirdly formatted in terms of line breaks, parantheses, etc.

File: domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+            {
+                HttpContext = new DefaultHttpContext()
+            }
+        };
+    }
+
+    [Fact]
+    public void ForwardAuth_OpaqueTokenAndInternalTokenExists_ReturnAuthorizationHeaderAndStatusCode200()
+    {
+        tokenStorage.Setup(x => x.GetInteralTokenByOpaqueToken(It.IsAny<string>())).Returns(new InternalToken
```
Comment (endk-awo) [inline domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:48]: Consider using:

// Arrange
// Act
// Assert

File: domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+        var response = tokenController.ForwardAuth();
+
+        Assert.Equal(200, ((StatusCodeResult)response).StatusCode);
+    }
+
+    [Fact]
+    public void ForwardAuth_MissingOpaqueTokenInHeader_ReturnStatusCode401()
+    {
+        var response = tokenController.ForwardAuth();
```
Comment (endk-awo) [inline domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:69]: Consider using:

// Arrange
// Act
// Assert

File: domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+        Assert.Equal(401, ((StatusCodeResult)response).StatusCode);
+    }
+
+    [Fact]
+    public void ForwardAuth_OpaqueTokenExistsInHeaderNoInternalToken_ReturnStatusCode401()
+    {
+        var tokenStorage = new Mock<ITokenStorage>();
```
Comment (endk-awo) [inline domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:77]: Consider using:

// Arrange
// Act
// Assert

File: domains/auth/Auth.API/Tests/Repository/TestTokenStorage.cs [hunk 17]
```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Collections.Generic;
+using API.Models;
+using API.TokenStorage;
+using Xunit;
+using Xunit.Categories;
+
+namespace Tests.Controller;
+
+
```
Comment (endk-awo) [inline domains/auth/Auth.API/Tests/Repository/TestTokenStorage.cs:10]: Redundant newline

File: domains/auth/Auth.API/Tests/Repository/TestTokenStorage.cs [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+    {
+            new object[] { DateTime.UtcNow.AddHours(-1), DateTime.UtcNow.AddHours(5), true },
+            new object[] { DateTime.UtcNow.AddHours(1), DateTime.UtcNow.AddHours(6), false },
+            new object[] { DateTime.UtcNow.AddHours(-10), DateTime.UtcNow.AddHours(-4), false },
+    };
+
+    [Theory, MemberData(nameof(CookieValidation))]
+    public void TokenStorage_InteralTokenIssuedAndExpiresValidation(DateTime issued, DateTime expires, bool expectedIsValid)
+    {
+        var tokenStorage = new TokenStorage();
```
Comment (endk-awo) [inline domains/auth/Auth.API/Tests/Repository/TestTokenStorage.cs:24]: Consider using:

// Arrange
// Act
// Assert

File: domains/auth/Auth.API/Tests/Repository/TestTokenStorage.cs [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+            Expires = expires,
+        };
+
+        var isValid = tokenStorage.InternalTokenValidation(internalToken);
+
+        Assert.Equal(expectedIsValid, isValid);
+    }
+
+    public static IEnumerable<object[]> InternalTokens => new[]
+{
```
Comment (endk-awo) [inline domains/auth/Auth.API/Tests/Repository/TestTokenStorage.cs:41]: Indentation

File: domains/auth/Auth.API/Tests/Repository/TestTokenStorage.cs [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+            new object[] { new InternalToken(){ Actor = "Actor", Subject = "Subject", Scope = new List<string> { "Scope1", "Scope2" }, Issued = DateTime.UtcNow.AddHours(-1), Expires = DateTime.UtcNow.AddHours(5) }, true },
+            new object[] { null , false},
+    };
+
+    [Theory, MemberData(nameof(InternalTokens))]
+    public void TokenStorage_InteralTokenIsNotNullValidation(InternalToken? interalToken, bool expectedIsValid)
+    {
+        var tokenStorage = new TokenStorage();
```
Comment (endk-awo) [inline domains/auth/Auth.API/Tests/Repository/TestTokenStorage.cs:49]: Consider using:

// Arrange
// Act
// Assert

**Review by endk-awo — CHANGES_REQUESTED** [review]

File: domains/auth/Auth.API/API/Controllers/TokenController.cs (hunk 2 above)
Comment (ckr123) [reply domains/auth/Auth.API/API/Controllers/TokenController.cs]: Could this be fixed within the editorconfig?

File: domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs (hunk 16 above)
Comment (ckr123) [reply domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:76]: I would actually prefer not to have comments in the code.

File: domains/auth/Auth.API/API/Utilities/Cookies.cs [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+namespace API.Services;
+
+public class Cookies : ICookies
```
Comment (ckr123) [inline domains/auth/Auth.API/API/Utilities/Cookies.cs:7]: @C-Christiansen was the only usage of this utility in the tokencontroller to create the test endpoint? Or did i forget something?

File: domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs (hunk 13 above)
Comment (ckr123) [reply domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:32]: How would you prefer it? I actually think it is quite easy to read when you quickly skim the file

File: domains/auth/Auth.API/API/Repository/TokenStorage.cs [hunk 22]
```diff
@@ -11,5 +13,40 @@ public string GetIdTokenByOpaqueToken(string token)
...
         }
+
+        public bool InternalTokenValidation(InternalToken interalToken)
+        {
+            if (interalToken == null)
+            {
+                return false;
+            }
+
+            if (interalToken.Issued > DateTime.UtcNow || interalToken.Expires < DateTime.UtcNow)
```
Comment (ckr123) [inline domains/auth/Auth.API/API/Repository/TokenStorage.cs]: @C-Christiansen Maybe we can extract this boolean to a private method tobe more descriptive on what it does?

File: domains/auth/Auth.API/API/Repository/TokenStorage.cs [hunk 23]
```diff
@@ -11,5 +13,40 @@ public string GetIdTokenByOpaqueToken(string token)
...
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        public InternalToken? GetInteralTokenByOpaqueToken(string token)
```
Comment (ckr123) [inline domains/auth/Auth.API/API/Repository/TokenStorage.cs]: @C-Christiansen I might have a suggestion to this method. Lets have a talk tomorrow 👍

File: domains/auth/Auth.API/API/Repository/TokenStorage.cs (hunk 22 above)
Comment (ckr123) [reply domains/auth/Auth.API/API/Repository/TokenStorage.cs]: Extracted to fluentvalidation

File: domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs (hunk 16 above)
Comment (endk-awo) [reply domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:74]: Okay, that's a different discussion then. But I find that *some* way of separating these steps in a unit test helps the immediate overview and readability of it.

File: domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs (hunk 13 above)
Comment (endk-awo) [reply domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:30]:
```
        tokenController = new (authOptionsMock.Object, tokenStorage.Object)
        {
            ControllerContext = new () { HttpContext = new DefaultHttpContext() }
        };
```

I'd do it something like this. I don't think the extra line breaks help the clarity when the lines can still be this short like this.

Comment (ckr123) [reply domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:30]: I think this should be something for the linter to take care of.
Untill we have a linting rule, lets not focus on this.

File: domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs (hunk 16 above)
Comment (ckr123) [reply domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:74]: We havent used it in this project at all, so i would prefer not to add it for now.

Comment (endk-awo) [reply domains/auth/Auth.API/Tests/Controllers/TestTokenController.cs:74]: If you look at the Emissions and Measurements API's, you will find them on most tests. I do think we should align our way of structuring tests.

**Review by endk-awo — APPROVED** [review]

**Pull request merged**